### ISSUE TITLE
Possibility to read/edit SpineDB when executed through external IDE

### ISSUE THREAD

**Issue by jimhommes** [Spine-project/Spine-Toolbox#1355]
In my current project I am reading and editing the database through my Python code, but this has to be launched from within the project. I was wondering whether it's possible to read and edit the SpineDB when code is executed externally through my IDE. I've tried passing the url that's passed through Spine, but that gives me

```
ConnectionRefusedError: [WinError 10061] No connection could be made because the target machine actively refused it
```

Is it possible? It would spare time overall, but it's only a nice to have.

**Comment by soininen** [comment]
The URL given through the Tool project item in Toolbox is in fact not the database's URL but a link to a special 'database server' that is created for the Tool by Toolbox. Its purpose is to prevent Tools that execute in parallel from messing up each other's database accesses.

Normally you would connect to a database using its direct URL, e.g. `sqlite:///c:\path\to\my\data\database.sqlite`

**Comment by manuelma** [comment]
It is certainly possible. What are you doing exactly to get that error?

**Comment by manuelma** [comment]
> Normally you would connect to a database using its direct URL, e.g. sqlite:///c:\path\to\my\data\database.sqlite

You can also call `[Diff]DatabaseMapping(url-of-the-server)`

**Comment by soininen** [comment]
> You can also call [Diff]DatabaseMapping(url-of-the-server)

Don't you then have to set up the server manually? That's my understanding but I might be wrong here.

**Comment by jimhommes** [comment]
> Normally you would connect to a database using its direct URL, e.g. `sqlite:///c:\path\to\my\data\database.sqlite`

Works like a charm! I assumed that Spine would be running the server and the localhost address provided by Spine would be necessary. Thanks!

**Comment by soininen** [comment]
It is recommended to use the localhost address provided by Toolbox when running a script in Toolbox. When executing outside of Toolbox e.g. in IDE or console the direct URL works instead as the database server is not available unless you set it up yourself.

**Comment by manuelma** [comment]
I'm still curious about what produced that error, can you @jimhommes provide more detail?

**Comment by manuelma** [comment]
> Don't you then have to set up the server manually? 

You do, I guess I didn't understand the question.

**Comment by jimhommes** [comment]
> I'm still curious about what produced that error, can you @jimhommes provide more detail?

@manuelma I use Python code to read and edit SpineDB through the provided spinedb.py module through
```
db_url = sys.argv[1]
db = SpineDB(db_url)
```

I tried to run it from the IDE, so I copied the address `http://localhost:50260` provided by Spine Toolbox into my IDE as argument. Running the IDE then gave that error. It works now using the direct path.

**Comment by PekkaSavolainen** [comment]
What is `spinedb.py`?

**Comment by manuelma** [comment]
Thanks @jimhommes that makes sense, the server only lives during execution so it's normal to get an error when trying to access it outside that context.

**Comment by jimhommes** [comment]
@PekkaSavolainen Python file from the `EMP-E-2-models-demo` project which calls the `spinedb_api` packages.

**Comment by manuelma** [comment]
I thought I'd reopen this jut to ask @jimhommes, why is it that you prefer to work on your IDE than in Spine Toolbox? I have experienced the same feeling, so do you have any ideas on how to improve Spine Toolbox so you wouldn't need your IDE?

**Comment by jimhommes** [comment]
The project I'm working on now has quite a large Python project attached to it (multiple files). The main reason why I'm using an IDE at all is because of the project structure and it's sophisticated syntax checking/highlighting. 

I've used the Spine IDE before, but mostly as quick scripting. I see is as insightful and useful but not for larger files/projects with interdependencies.

**Comment by manuelma** [comment]
Thanks very much @jimhommes . You're right, Spine Toolbox is not designed for that use case, so we're cool. It looks like the headless mode as suggested by @soininen could be used for IDE integration, and of course you can try that out if you're interested.
